src/core/ucc_lib.c

```diff
@@ -19,13 +21,235 @@ static ucc_config_field_t ucc_lib_config_table[] = {
+        cl_lib->priority = (-1 == cl_config->priority) ? cl_iface->priority
+                                                       : cl_config->priority;
+        ucc_config_parser_release_opts(cl_config,
+                                       cl_iface->cl_lib_config.table);
+        free(cl_config);
+        lib->libs[lib->n_libs_opened++] = cl_lib;
+        supported_coll_types |= cl_iface->attr.coll_types;
+        if (cl_iface->attr.thread_mode < supported_tm) {
+            supported_tm = cl_iface->attr.thread_mode;
+        }
+        ucc_info("lib_prefix \"%s\": initialized component \"%s\" priority %d",
+                 config->full_prefix, cl_iface->super.name, cl_lib->priority);
+    }
+
+    /* Check if the combination of the selected CLs provides all the
+       requested coll_types: not an error, just print a message if not
+       all the colls are supproted */
+    if (params.mask & UCC_LIB_PARAM_FIELD_COLL_TYPES &&
+        ((params.coll_types & supported_coll_types) != params.coll_types)) {
+        ucc_debug("selected set of CLs does not provide all the requested "
+                  "coll_types");
+    }
+    if (params.thread_mode > supported_tm) {
+        ucc_debug("selected set of CLs does not provide the requested "
+                  "thread_mode");
+    }
+    lib->attr.coll_types  = supported_coll_types;
+    lib->attr.thread_mode = ucc_min(supported_tm, params.thread_mode);
+    return UCS_OK;
+
+error_cl_init:
+    ucc_config_parser_release_opts(cl_config, cl_iface->cl_lib_config.table);
+error_cl_config_parse:
+    free(cl_config);
+error_cl_cleanup:
+    for (i = 0; i < lib->n_libs_opened; i++) {
+        lib->libs[i]->iface->finalize(lib->libs[i]);
+    }
+error:
+    if (lib->libs)
+        free(lib->libs);
+    return status;
+}
+
 ucc_status_t ucc_init_version(unsigned api_major_version,
                               unsigned api_minor_version,
                               const ucc_lib_params_t *params,
-                              const ucc_lib_config_h *config,
-                              ucc_lib_h *lib_p)
+                              const ucc_lib_config_h config, ucc_lib_h *lib_p)
 {
-    return UCC_ERR_NOT_IMPLEMENTED;
+    unsigned        major_version, minor_version, release_number;
+    ucc_status_t    status;
+    ucc_lib_info_t *lib;
+
+    *lib_p = NULL;
+
+    if (UCC_OK != (status = ucc_constructor())) {
+        return status;
+    }
+
+    ucc_get_version(&major_version, &minor_version, &release_number);
+
+    if ((api_major_version != major_version) ||
+        ((api_major_version == major_version) &&
+         (api_minor_version > minor_version))) {
+        ucc_warn(
+            "UCC version is incompatible, required: %d.%d, actual: %d.%d.%d",
+            api_major_version, api_minor_version, major_version, minor_version,
+            release_number);
+    }
+
+    if (0 == ucc_global_config.cl_framework.n_components) {
```

Comment:
i think it is initialized in ucc_component.c: 81 when we call ucc_componend_load from ucc_constructor. however i agree the check is not needed, since we check constructor return.